kernel/src/vm/vmar/mod.rs

```diff
@@ -277,6 +326,65 @@ impl VmarInner {
 
         return_errno_with_message!(Errno::ENOMEM, "Cannot find free region for mapping");
     }
+
+    /// Splits and unmaps the found mapping if the new size is smaller.
+    /// Enlarges the last mapping if the new size is larger.
+    fn resize_mapping(
+        &mut self,
+        vm_space: &VmSpace,
+        map_addr: Vaddr,
+        old_size: usize,
+        new_size: usize,
+        rss_delta: &mut RssDelta,
+    ) -> Result<()> {
+        debug_assert!(map_addr % PAGE_SIZE == 0);
+        debug_assert!(old_size % PAGE_SIZE == 0);
+        debug_assert!(new_size % PAGE_SIZE == 0);
+
+        // FIXME: We should check whether all existing ranges in
+        // `map_addr..map_addr + old_size` have a mapping. If not,
+        // we should return a `Err`.
+
+        if new_size == 0 {
+            return_errno_with_message!(Errno::EINVAL, "can not resize a mapping to 0 size");
+        }
+
+        if new_size == old_size {
+            return Ok(());
+        }
+
+        let old_map_end = map_addr + old_size;
+        let new_map_end = map_addr + new_size;
```

Comment:
- For the old mappings, it shouldn't overflow as it's something that the caller guarantees (e.g., via `check_lies_in_single_mapping`).
 - For the new mappings, maybe we need to add some checks (`checked_add` & `is_userspace_vaddr`) similar to `fn remap`.